ChainSharp/Extensions/WorkflowExtensions.cs

```
@@ -0,0 +1,146 @@
+    {
+        T? input = default;
+
+        var inputType = typeof(T);
+        if (inputType.IsTuple())
+        {
+            try
+            {
+                input = ExtractTuple(workflow, inputType);
+            }
+            catch (Exception e)
+            {
+                workflow.Exception ??= e;
+            }
+        }
+
+        input ??= (T?)workflow.Memory.GetValueOrDefault(inputType);
+        
+        if (input is null)
+            workflow.Exception ??= new WorkflowException($"Could not find type: ({inputType}).");
+
+        return input;
+    }
+
+    internal static dynamic[] ExtractTypesFromMemory<TInput, TReturn>(this Workflow<TInput, TReturn> workflow, IEnumerable<Type> types)
+        => types.Select(type => ExtractTypeFromMemory(workflow, type)).ToArray();
+    
+    internal static dynamic ExtractTypeFromMemory<TInput, TReturn>(this Workflow<TInput, TReturn> workflow, Type tIn)
+    {
+        dynamic? input = null;
+
+        var inputType = tIn;
+        if (inputType.IsTuple())
+        {
+            try
+            {
+                input = ExtractTuple(workflow, inputType);
+            }
+            catch (Exception e)
+            {
+                workflow.Exception ??= e;
+            }
+        }
+
+        input ??= workflow.Memory.GetValueOrDefault(inputType);
+        
+        if (input is null)
+            workflow.Exception ??= new WorkflowException($"Could not find type: ({inputType}).");
+
+        return input!;
+    }
+    
+    internal static dynamic ExtractTuple<TInput, TReturn>(this Workflow<TInput, TReturn> workflow, Type inputType)
+    {
+        var typeTuples = TypeHelpers.ExtractTypeTuples(workflow.Memory, inputType);
+
+        return typeTuples.Count switch
+        {
+            0 => throw new WorkflowException($"Cannot have Tuple of length 0."),
+            2 => TypeHelpers.ConvertTwoTuple(typeTuples),
```

I wonder if we should throw an exception if someone passes us a 1 item tuple. Tell them to just use the type itself!

---

Good idea -- Will add this as an issue